feat(signing): Add on-chain address signing & verification support

Anish-ai opened the pull request:
Extend message signing and verification to support on-chain address keys in addition to Lightning node pubkeys. This implementation:

- Add toggle between Lightning and on-chain signature modes
- Integrate with LND's SignMessageWithAddr endpoint for on-chain keys
- Implement address selection dropdown with account information
- Add sorting functionality (by balance or creation time)
- Fix address sorting logic to properly handle derivation paths
- Enhance UI with informative sections, visual separators, and better feedback
- Improve error handling and verification result display

This feature allows users to prove ownership of their Bitcoin addresses without spending funds, enhancing verification use cases like proving address ownership for services or exchanges.

# Description

Relates to issue: **ZEUS-0000**

_Please enter a description and screenshots, if appropriate, of the work covered in this PR_

This pull request is categorized as a:

- [x] New feature
- [ ] Bug fix
- [ ] Code refactor
- [ ] Configuration change
- [ ] Locales update
- [ ] Quality assurance
- [ ] Other

## Checklist
- [x] I’ve run `yarn run tsc` and made sure my code compiles correctly
- [x] I’ve run `yarn run lint` and made sure my code didn’t contain any problematic patterns
- [ ] I’ve run `yarn run prettier` and made sure my code is formatted correctly
- [x] I’ve run `yarn run test` and made sure all of the tests pass

## Testing

If you modified or added a utility file, did you add new unit tests?

- [ ] No, I’m a fool
- [ ] Yes
- [ ] N/A

I have tested this PR on the following platforms (please specify OS version and phone model/VM):

- [x] Android
- [ ] iOS

I have tested this PR with the following types of nodes (please specify node version and API version where appropriate):

- [x] Embedded LND
- [x] LND (REST)
- [x] LND (Lightning Node Connect)
- [x] Core Lightning (CLNRest)
- [ ] Nostr Wallet Connect
- [ ] LndHub

### Locales
- [ ] I’ve added new locale text that requires translations
- [ ] I’m aware that new translations should be made on the ZEUS 

### Third Party Dependencies and Packages

- [ ] Contributors will need to run `yarn` after this PR is merged in
- [ ] 3rd party dependencies have been modified:
    * verify that `package.json` and `yarn.lock` have been properly updated
    * verify that dependencies are installed for both iOS and Android platforms

### Other:

- [ ] Changes were made that require an update to the README
- [ ] Changes were made that require an update to onboarding


https://github.com/user-attachments/assets/5ec96735-9c0d-4e69-82f8-31b63eb7223e

kaloudis commented:
nit: rebases are preferred to merge commits

Anish-ai commented:
> nit: rebases are preferred to merge commits

Rebased the branch as suggested, thanks for pointing it out!

kaloudis commented on locales/en.json:
```diff
@@ -762,6 +762,25 @@
+    "views.Settings.SignMessage.loadingAddresses": "Loading addresses...",
+    "views.Settings.SignMessage.fullAddress": "Full Address:",
+    "views.Settings.SignMessage.account": "Account:",
+    "views.Settings.SignMessage.addressType": "Address Type:",
```
don't include colons, `:` in the locale strings. Place them in the view

Anish-ai replied:
> don't include colons, `:` in the locale strings. Place them in the view

Removed colons from locale strings and used the existing addressType key in the view. Thanks for the catch! Let me know if anything else needs tweaking 🙌

kaloudis commented on locales/en.json:
```diff
@@ -762,6 +762,25 @@
+    "views.Settings.SignMessage.loadingAddresses": "Loading addresses...",
+    "views.Settings.SignMessage.fullAddress": "Full Address:",
+    "views.Settings.SignMessage.account": "Account:",
+    "views.Settings.SignMessage.addressType": "Address Type:",
```
We already have a string for address type

kaloudis commented:
The commit structure could use some work. We also prefer rebases to merge commits.

Anish-ai commented:
> The commit structure could use some work. We also prefer rebases to merge commits.

Rebased and cleaned up the commits. Squashed all of them into one.

Anish-ai commented:
@kaloudis Hi! This is ready for review. I’m continuously rebasing it with the upstream master to keep it up to date. Let me know if any changes are needed.

kaloudis commented on App.tsx:
```diff
@@ -284,7 +284,7 @@ export default class App extends React.PureComponent {
                                                     if (nav != null) {
                                                         NavigationService.setTopLevelNavigator(
-                                                            nav
+                                                            nav as any
```
why was this change made? it doesn't throw any errors in our test runners right now. If we add types we should add specific ones, not any

kaloudis commented on yarn.lock:
why is this updated? there are not changes to our packages

kaloudis commented on views/Tools/SignVerifyMessage.tsx:
```diff
@@ -30,24 +35,366 @@ interface SignVerifyMessageState {
+}
+
+// Convert the static styles to a theme-aware function
+const getStyles = (theme: any) =>
```
this doesn't match our code style

kaloudis commented on views/Tools/SignVerifyMessage.tsx:
```diff
@@ -69,29 +419,410 @@ export default class SignVerifyMessage extends React.Component<
+                <View style={this.styles.form}>
+                    <Text style={{ color: themeColor('secondaryText') }}>
+                        Only Lightning Node message signing is available in this
+                        implementation.
```
this should be localized

kaloudis commented on views/Tools/SignVerifyMessage.tsx:
```diff
@@ -69,29 +419,410 @@ export default class SignVerifyMessage extends React.Component<
+                                        MessageSignStore.setSelectedAddress(
+                                            value
+                                        );
+                                        // Adding console log to debug
```
why do we need a comment for a logger?

kaloudis commented on views/Tools/SignVerifyMessage.tsx:
```diff
@@ -69,29 +419,410 @@ export default class SignVerifyMessage extends React.Component<
+                                                    (this.state
+                                                        .signingAddressLocal ||
+                                                        selectedAddress)
+                                            )?.type || 'Unknown'}
```
Unknown should be localized

kaloudis commented on views/Tools/SignVerifyMessage.tsx:
```diff
@@ -135,254 +868,753 @@ export default class SignVerifyMessage extends React.Component<
+                                {selectedIndex === 0
+                                    ? 'Sign a message to prove ownership of your Bitcoin address or Lightning node. This creates a cryptographic signature that others can verify.'
+                                    : 'Verify a cryptographic signature to confirm a message was signed by a specific Bitcoin address or Lightning node.'}
+                            </Text>
```
these strings should be localized

kaloudis commented on views/Tools/SignVerifyMessage.tsx:
```diff
@@ -135,254 +868,753 @@ export default class SignVerifyMessage extends React.Component<
+                                        Tip: Signing messages can be used to
+                                        prove ownership without revealing
+                                        private keys or spending funds.
+                                    </Text>
```
localize

kaloudis commented on views/Tools/SignVerifyMessage.tsx:
```diff
@@ -135,254 +868,753 @@ export default class SignVerifyMessage extends React.Component<
-                                        </Text>
+                                            <CopyButton
+                                                copyValue={signature}
+                                                title="Copy Signature"
```
localize

kaloudis commented on views/Tools/SignVerifyMessage.tsx:
```diff
@@ -135,254 +868,753 @@ export default class SignVerifyMessage extends React.Component<
+                                                }
+                                                buttons={[
+                                                    'Lightning Node',
+                                                    'On-chain Address'
```
these should be localized

kaloudis commented on views/Tools/SignVerifyMessage.tsx:
```diff
@@ -135,254 +868,753 @@ export default class SignVerifyMessage extends React.Component<
+                                                )}
+                                            </Text>
+
+                                            {/* Add sorting dropdown for verify section */}
```
we do need this comment - seems AI generated

kaloudis commented on views/Tools/SignVerifyMessage.tsx:
```diff
@@ -135,254 +868,753 @@ export default class SignVerifyMessage extends React.Component<
+                                                        No on-chain addresses
+                                                        available. Please ensure
+                                                        your node supports
+                                                        address management.
```
localize

kaloudis commented on views/Tools/SignVerifyMessage.tsx:
```diff
@@ -135,254 +868,753 @@ export default class SignVerifyMessage extends React.Component<
+                                        <View style={this.styles.infoCard}>
+                                            <Text style={this.styles.infoText}>
+                                                Lightning node verification
+                                                selected. No address needed.
```
localize - but why is the address picker even shown when in lightning node signing?

kaloudis commented on views/Tools/SignVerifyMessage.tsx:
```diff
@@ -135,254 +868,753 @@ export default class SignVerifyMessage extends React.Component<
+                                            ? 'Lightning node'
+                                            : 'Bitcoin address'}{' '}
+                                        without requiring them to reveal private
+                                        keys.
```
localize

kaloudis commented on views/Tools/SignVerifyMessage.tsx:
```diff
@@ -135,254 +868,753 @@ export default class SignVerifyMessage extends React.Component<
+                                                    {signingMode === 'lightning'
+                                                        ? 'Lightning node'
+                                                        : 'Bitcoin address'}{' '}
+                                                    owner.
```
localize

kaloudis commented on views/Tools/SignVerifyMessage.tsx:
```diff
@@ -135,254 +868,753 @@ export default class SignVerifyMessage extends React.Component<
+                                                    {signingMode === 'lightning'
+                                                        ? 'Lightning node'
+                                                        : 'Bitcoin address'}
+                                                    .
```
localize

kaloudis commented:
The UI needs a ton of work.

- A lot of the styles don't match our existing styles.
- The explainer on the top of the page is confusing and out of place (better suited for a separated page or pop up modal).
- Tips should me modals or other pages.
- Header style doesn't match any of our existing views.
- Address picker is quite rudimentary. It won't scale if the user has a bunch of addresses. We need some sort of dedicated page for this where you can sort and filter.

Anish-ai commented:
> The UI needs a ton of work.
> 
> - A lot of the styles don't match our existing styles.
> - The explainer on the top of the page is confusing and out of place (better suited for a separated page or pop up modal).
> - Tips should me modals or other pages.
> - Header style doesn't match any of our existing views.
> - Address picker is quite rudimentary. It won't scale if the user has a bunch of addresses. We need some sort of dedicated page for this where you can sort and filter.

Thanks for the feedback! I’ll start working on the UI fixes and improvements ASAP and push the updates soon.

Anish-ai commented:
@kaloudis Quick question: While reworking the address picker UI, should I also add options like “hide zero balance addresses” and possibly various sorting/filtering capabilities (like on the on-chain addresses page) to maintain consistency?

kaloudis commented:
Closing in favor of https://github.com/ZeusLN/zeus/pull/2945

@Anish-ai some general feedback:
- Learn to use branches instead of submitting PRs from the `master` branch of your fork
- Rebases are generally preferred over merge commits
